### PR TITLE
Drop environment columns from core tables

### LINKED ISSUE #270: Plugin uninstall fails on Host having an Environment set
`foreman-maintain plugin purge-puppet` uses a down migration for uninstalling puppet plugin, this unfortunatelly fails on Host referring to environments table for hosts, that have environment set.

```
[root@dhcp-3-32 ~]# foreman-maintain plugin purge-puppet
Running Remove Puppet feature
================================================================================
Check for Puppet capsules from the database:                          [OK]
--------------------------------------------------------------------------------
Remove Puppet feature:                                                [FAIL]
Failed executing foreman-rake db:migrate VERSION=0 SCOPE=foreman_puppet, exit status 1:
 'ErbParser' is ignored.
'RubyParser' is ignored.
rake aborted!
StandardError: An error has occurred, this and all later migrations canceled:
 
PG::DependentObjectsStillExist: ERROR:  cannot drop table environments because other objects depend on it
DETAIL:  constraint hostgroups_environment_id_fk on table hostgroups depends on table environments
constraint hosts_environment_id_fk on table hosts depends on table environments
HINT:  Use DROP ... CASCADE to drop the dependent objects too.
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/connection_adapters/postgresql/database_statements.rb:92:in `async_exec'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/connection_adapters/postgresql/database_statements.rb:92:in `block (2 levels) in execute'
/opt/theforeman/tfm/root/usr/share/gems/gems/activesupport-6.0.3.7/lib/active_support/dependencies/interlock.rb:48:in `block in permit_concurrent_loads'
/opt/theforeman/tfm/root/usr/share/gems/gems/activesupport-6.0.3.7/lib/active_support/concurrency/share_lock.rb:187:in `yield_shares'
/opt/theforeman/tfm/root/usr/share/gems/gems/activesupport-6.0.3.7/lib/active_support/dependencies/interlock.rb:47:in `permit_concurrent_loads'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/connection_adapters/postgresql/database_statements.rb:91:in `block in execute'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/connection_adapters/abstract_adapter.rb:722:in `block (2 levels) in log'
/opt/theforeman/tfm/root/usr/share/gems/gems/activesupport-6.0.3.7/lib/active_support/concurrency/load_interlock_aware_monitor.rb:26:in `block (2 levels) in synchronize'
/opt/theforeman/tfm/root/usr/share/gems/gems/activesupport-6.0.3.7/lib/active_support/concurrency/load_interlock_aware_monitor.rb:25:in `handle_interrupt'
/opt/theforeman/tfm/root/usr/share/gems/gems/activesupport-6.0.3.7/lib/active_support/concurrency/load_interlock_aware_monitor.rb:25:in `block in synchronize'
/opt/theforeman/tfm/root/usr/share/gems/gems/activesupport-6.0.3.7/lib/active_support/concurrency/load_interlock_aware_monitor.rb:21:in `handle_interrupt'
/opt/theforeman/tfm/root/usr/share/gems/gems/activesupport-6.0.3.7/lib/active_support/concurrency/load_interlock_aware_monitor.rb:21:in `synchronize'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/connection_adapters/abstract_adapter.rb:721:in `block in log'
/opt/theforeman/tfm/root/usr/share/gems/gems/activesupport-6.0.3.7/lib/active_support/notifications/instrumenter.rb:24:in `instrument'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/connection_adapters/abstract_adapter.rb:712:in `log'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/connection_adapters/postgresql/database_statements.rb:90:in `execute'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/connection_adapters/postgresql/schema_statements.rb:58:in `drop_table'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/migration.rb:890:in `block in method_missing'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/migration.rb:858:in `block in say_with_time'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/migration.rb:858:in `say_with_time'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/migration.rb:879:in `method_missing'
/opt/theforeman/tfm/root/usr/share/gems/gems/foreman_puppet-2.0.5/db/migrate_foreman/20090722141107_create_environments.foreman_puppet.rb:15:in `down'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/migration.rb:831:in `exec_migration'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/migration.rb:812:in `block (2 levels) in migrate'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/migration.rb:811:in `block in migrate'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/connection_adapters/abstract/connection_pool.rb:471:in `with_connection'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/migration.rb:810:in `migrate'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/migration.rb:1002:in `migrate'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/migration.rb:1310:in `block in execute_migration_in_transaction'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/migration.rb:1361:in `block in ddl_transaction'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/connection_adapters/abstract/database_statements.rb:280:in `block in transaction'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/connection_adapters/abstract/transaction.rb:280:in `block in within_new_transaction'
/opt/theforeman/tfm/root/usr/share/gems/gems/activesupport-6.0.3.7/lib/active_support/concurrency/load_interlock_aware_monitor.rb:26:in `block (2 levels) in synchronize'
/opt/theforeman/tfm/root/usr/share/gems/gems/activesupport-6.0.3.7/lib/active_support/concurrency/load_interlock_aware_monitor.rb:25:in `handle_interrupt'
/opt/theforeman/tfm/root/usr/share/gems/gems/activesupport-6.0.3.7/lib/active_support/concurrency/load_interlock_aware_monitor.rb:25:in `block in synchronize'
/opt/theforeman/tfm/root/usr/share/gems/gems/activesupport-6.0.3.7/lib/active_support/concurrency/load_interlock_aware_monitor.rb:21:in `handle_interrupt'
/opt/theforeman/tfm/root/usr/share/gems/gems/activesupport-6.0.3.7/lib/active_support/concurrency/load_interlock_aware_monitor.rb:21:in `synchronize'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/connection_adapters/abstract/transaction.rb:278:in `within_new_transaction'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/connection_adapters/abstract/database_statements.rb:280:in `transaction'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/transactions.rb:212:in `transaction'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/migration.rb:1361:in `ddl_transaction'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/migration.rb:1309:in `execute_migration_in_transaction'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/migration.rb:1281:in `block in migrate_without_lock'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/migration.rb:1280:in `each'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/migration.rb:1280:in `migrate_without_lock'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/migration.rb:1229:in `block in migrate'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/migration.rb:1382:in `with_advisory_lock'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/migration.rb:1229:in `migrate'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/migration.rb:1071:in `down'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/migration.rb:1040:in `migrate'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/tasks/database_tasks.rb:238:in `migrate'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/railties/databases.rake:86:in `block (3 levels) in <top (required)>'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/railties/databases.rake:84:in `each'
/opt/theforeman/tfm/root/usr/share/gems/gems/activerecord-6.0.3.7/lib/active_record/railties/databases.rake:84:in `block (2 levels) in <top (required)>'
/opt/rh/rh-ruby27/root/usr/share/gems/gems/rake-13.0.1/exe/rake:27:in `<top (required)>'
```

### PR DESCRIPTION
**Pull request by ezr-ondrej**
Fixes #270

### CODE REVIEW

File: db/migrate/20220421204325_drop_environment_from_host_and_hg.foreman_puppet.rb [hunk 1]
```diff
@@ -0,0 +1,8 @@
+class DropEnvironmentEnvironmentFromHostAndHg < ActiveRecord::Migration[6.0]
```
Comment (nadjaheitmann) [inline db/migrate/20220421204325_drop_environment_from_host_and_hg.foreman_puppet.rb:1]:
```suggestion
class DropEnvironmentFromHostAndHostgroup < ActiveRecord::Migration[6.0]
```

File: db/migrate/20220421204325_drop_environment_from_host_and_hg.foreman_puppet.rb [hunk 2]
```diff
@@ -0,0 +1,8 @@
...
+  def up
```
Comment (nadjaheitmann) [inline db/migrate/20220421204325_drop_environment_from_host_and_hg.foreman_puppet.rb:2]:
```suggestion
  def down
```

**Review by nadjaheitmann — CHANGES_REQUESTED** [review]
If you change the name of the migration, you need also change the file name I guess :)

**Comment by nadjaheitmann** [comment]
From my understanding of Rails, this migration would run as one of the first ones. Don't you think that some other tables need to be cleaned up, first?

**Comment by nadjaheitmann** [comment]
I just found this in the migrations: https://github.com/theforeman/foreman_puppet/blob/master/db/migrate/20211112130803_cleanup_environment_from_core_tables.foreman_puppet.rb

Maybe your code could also be part of this migration.

**Comment by nadjaheitmann** [comment]
Also, from my understanding, this will not yet solve #271 because the error does not occur on installation but on uninstallation, so we need a down migration to solve this?

**Comment by ezr-ondrej** [comment]
> I just found this in the migrations: https://github.com/theforeman/foreman_puppet/blob/master/db/migrate/20211112130803_cleanup_environment_from_core_tables.foreman_puppet.rb

You're right, seeing this, that should prevent the error on uninstallations :thinking:

**Comment by ezr-ondrej** [comment]
> because the error does not occur on installation but on uninstallation, so we need a down migration to solve this?

This would remove the column on up migrations, so there would be no issue on down migrations.

**Comment by nadjaheitmann** [comment]
> This would remove the column on up migrations, so there would be no issue on down migrations (uninstall) with this column.

The guy who had the problems of removing the Puppet data never installed the plugin in the first place, just foreman core. So any up migrations we add to the plugin code are not run for people who are just dropping the data they don't need. While upgrading from 2.5, you can choose whether to install the plugin or not.

**Comment by nadjaheitmann** [comment]
> The guy who had the problems of removing the Puppet data never installed the plugin in the first place, just foreman core. So any up migrations we add to the plugin code are not run for people who are just dropping the data they don't need. While upgrading from 2.5, you can choose whether to install the plugin or not.

I see that he did via foreman-installer, you don't on the manual upgrade. So when you downgrade, eventually this is run:
https://github.com/theforeman/foreman_puppet/blob/master/db/migrate/20200803113803_migrate_environment_to_puppet_facet.foreman_puppet.rb

It moves the data back in the original db tables because the column still exists, it has not yet been dropped anywhere.

**Comment by ezr-ondrej** [comment]
> It moves the data back in the original db tables because the column still exists, it has not yet been dropped anywhere.

Right, so this should be the best path forward afterall

**Comment by ezr-ondrej** [comment]
Test fix: https://github.com/theforeman/foreman_puppet/pull/272

**Review by nadjaheitmann — APPROVED** [review]

**Comment by nadjaheitmann** [comment]
CP 2.0: 8401180257a88c7442c1ef82816ca0db396c39db

**Pull request merged**